File: VoiceControlApp/NetworkSettings.swift
import Foundation

// MARK: - Network Target Types

/// Defines where RCP commands should be sent
enum NetworkTargetType: String, CaseIterable {
    case console = "console"
    case testing = "testing"
    case both = "both"
    
    var displayName: String {
        switch self {
        case .console:
            return "Yamaha Console"
        case .testing:
            return "Mac GUI (Testing)"
        case .both:
            return "Console + GUI"
        }
    }
    
    var description: String {
        switch self {
        case .console:
            return "Send commands directly to Yamaha mixing console"
        case .testing:
            return "Send commands to Mac GUI receiver for development"
        case .both:
            return "Send to console and show in Mac GUI for monitoring"
        }
    }
}

// MARK: - Network Settings Model

/// Configuration settings for RCP command network transmission
class NetworkSettings: ObservableObject {
    
    // MARK: - Published Properties
    
    /// Current target type for RCP commands
    @Published var targetType: NetworkTargetType = .testing
    
    /// Yamaha console IP address
    @Published var consoleIP: String = "192.168.1.100"
    
    /// Yamaha console port (standard RCP port)
    @Published var consolePort: Int = 49280
    
    /// Mac GUI receiver IP address  
    @Published var testingIP: String = "192.168.1.50"
    
    /// Mac GUI receiver port
    @Published var testingPort: Int = 8080
    
    /// Connection timeout in seconds
    @Published var timeoutSeconds: Double = 5.0
    
    /// Enable detailed network logging
    @Published var enableLogging: Bool = true
    
    /// Last successful connection timestamp
    @Published var lastConnectionTime: Date?
    
    /// Current connection status
    @Published var connectionStatus: ConnectionStatus = .disconnected
    
    // MARK: - Singleton
    
    static let shared = NetworkSettings()
    
    // MARK: - UserDefaults Keys
    
    private enum Keys {
        static let targetType = "networkTargetType"
        static let consoleIP = "consoleIP"
        static let consolePort = "consolePort"
        static let testingIP = "testingIP"
        static let testingPort = "testingPort"
        static let timeoutSeconds = "networkTimeoutSeconds"
        static let enableLogging = "networkEnableLogging"
        static let lastConnectionTime = "lastConnectionTime"
    }
    
    // MARK: - Initialization
    
    private init() {
        loadSettings()
    }
    
    // MARK: - Computed Properties
    
    /// Get the primary target IP based on current settings
    var primaryTargetIP: String {
        switch targetType {
        case .console, .both:
            return consoleIP
        case .testing:
            return testingIP
        }
    }
    
    /// Get the primary target port based on current settings
    var primaryTargetPort: Int {
        switch targetType {
        case .console, .both:
            return consolePort
        case .testing:
            return testingPort
        }
    }
    
    /// Check if console communication is enabled
    var shouldSendToConsole: Bool {
        return targetType == .console || targetType == .both
    }
    
    /// Check if GUI testing communication is enabled
    var shouldSendToGUI: Bool {
        return targetType == .testing || targetType == .both
    }
    
    /// Validate current network configuration
    var isValidConfiguration: Bool {
        let consoleValid = isValidIP(consoleIP) && isValidPort(consolePort)
        let testingValid = isValidIP(testingIP) && isValidPort(testingPort)
        
        switch targetType {
        case .console:
            return consoleValid
        case .testing:
            return testingValid
        case .both:
            return consoleValid && testingValid
        }
    }
    
    // MARK: - Settings Persistence
    
    /// Load settings from UserDefaults
    private func loadSettings() {
        let defaults = UserDefaults.standard
        
        if let targetTypeString = defaults.object(forKey: Keys.targetType) as? String,
           let targetType = NetworkTargetType(rawValue: targetTypeString) {
            self.targetType = targetType
        }
        
        if let consoleIP = defaults.object(forKey: Keys.consoleIP) as? String {
            self.consoleIP = consoleIP
        }
        
        consolePort = defaults.object(forKey: Keys.consolePort) as? Int ?? 49280
        
        if let testingIP = defaults.object(forKey: Keys.testingIP) as? String {
            self.testingIP = testingIP
        }
        
        testingPort = defaults.object(forKey: Keys.testingPort) as? Int ?? 8080
        timeoutSeconds = defaults.object(forKey: Keys.timeoutSeconds) as? Double ?? 5.0
        enableLogging = defaults.object(forKey: Keys.enableLogging) as? Bool ?? true
        
        if let lastConnectionData = defaults.object(forKey: Keys.lastConnectionTime) as? Data {
            lastConnectionTime = try? JSONDecoder().decode(Date.self, from: lastConnectionData)
        }
    }
    
    /// Save settings to UserDefaults
    func saveSettings() {
        let defaults = UserDefaults.standard
        
        defaults.set(targetType.rawValue, forKey: Keys.targetType)
        defaults.set(consoleIP, forKey: Keys.consoleIP)
        defaults.set(consolePort, forKey: Keys.consolePort)
        defaults.set(testingIP, forKey: Keys.testingIP)
        defaults.set(testingPort, forKey: Keys.testingPort)
        defaults.set(timeoutSeconds, forKey: Keys.timeoutSeconds)
        defaults.set(enableLogging, forKey: Keys.enableLogging)
        
        if let lastConnectionTime = lastConnectionTime,
           let data = try? JSONEncoder().encode(lastConnectionTime) {
            defaults.set(data, forKey: Keys.lastConnectionTime)
        }
        
        defaults.synchronize()
        
        if enableLogging {
            print("📝 Network settings saved - Target: \(targetType.displayName)")
        }
    }
    
    // MARK: - Connection Status Management
    
    /// Update connection status and timestamp
    func updateConnectionStatus(_ status: ConnectionStatus) {
        connectionStatus = status
        
        if case .connected = status {
            lastConnectionTime = Date()
            saveSettings()
        }
        
        if enableLogging {
            print("🌐 Connection status: \(status)")
        }
    }
    
    // MARK: - Validation Helpers
    
    /// Validate IP address format
    private func isValidIP(_ ip: String) -> Bool {
        let parts = ip.split(separator: ".")
        guard parts.count == 4 else { return false }
        
        return parts.allSatisfy { part in
            guard let num = Int(part), num >= 0, num <= 255 else { return false }
            return true
        }
    }
    
    /// Validate port number
    private func isValidPort(_ port: Int) -> Bool {
        return port > 0 && port <= 65535
    }
    
    // MARK: - Factory Defaults
    
    /// Reset to factory default settings
    func resetToDefaults() {
        targetType = .testing
        consoleIP = "192.168.1.100"
        consolePort = 49280
        testingIP = "192.168.1.50"
        testingPort = 8080
        timeoutSeconds = 5.0
        enableLogging = true
        lastConnectionTime = nil
        connectionStatus = .disconnected
        
        saveSettings()
        
        print("🔄 Network settings reset to defaults")
    }
}

// MARK: - Connection Status

/// Represents the current network connection status
enum ConnectionStatus: Equatable {
    case disconnected
    case connecting
    case connected(String) // Connected with target info
    case error(String)     // Error with description
    
    var displayText: String {
        switch self {
        case .disconnected:
            return "Disconnected"
        case .connecting:
            return "Connecting..."
        case .connected(let target):
            return "Connected to \(target)"
        case .error(let message):
            return "Error: \(message)"
        }
    }
    
    var isConnected: Bool {
        if case .connected = self {
            return true
        }
        return false
    }
}

// MARK: - Preview Helper

#if DEBUG
extension NetworkSettings {
    /// Create settings instance for SwiftUI previews
    static var preview: NetworkSettings {
        let settings = NetworkSettings()
        settings.targetType = .testing
        settings.consoleIP = "192.168.1.100"
        settings.testingIP = "192.168.1.50"
        settings.connectionStatus = .connected("Mac GUI")
        return settings
    }
}
#endif

File: AndroidVoiceControlApp/app/src/main/java/com/voicecontrol/app/network/model/NetworkSettings.kt
package com.voicecontrol.app.network.model

import android.content.SharedPreferences
import android.util.Log
import com.voicecontrol.app.BuildConfig
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.Serializable
import javax.inject.Inject
import javax.inject.Named
import javax.inject.Singleton

/**
 * Network Settings Model for Voice Control App
 * 
 * Direct port of iOS NetworkSettings to Android with Kotlin
 * Manages network configuration for RCP communication with mixers
 * 
 * Features:
 * - IP address and port configuration for console and testing
 * - Target selection (console vs. testing GUI)
 * - Connection timeout and retry settings
 * - Network logging preferences
 * - Settings persistence with SharedPreferences
 * - Reactive state updates with StateFlow
 */
@Singleton
class NetworkSettings @Inject constructor(
    @Named("settings") private val sharedPreferences: SharedPreferences
) {
    
    companion object {
        private const val TAG = "NetworkSettings"
        
        // SharedPreferences keys
        private const val PREF_CONSOLE_IP = "network_console_ip"
        private const val PREF_CONSOLE_PORT = "network_console_port"
        private const val PREF_TESTING_IP = "network_testing_ip"
        private const val PREF_TESTING_PORT = "network_testing_port"
        private const val PREF_TARGET_TYPE = "network_target_type"
        private const val PREF_CONNECTION_STATUS = "network_connection_status"
        private const val PREF_TIMEOUT_SECONDS = "network_timeout_seconds"
        private const val PREF_ENABLE_LOGGING = "network_enable_logging"
        private const val PREF_LAST_CONNECTION_TIME = "network_last_connection_time"
        
        // Default values
        private const val DEFAULT_CONSOLE_IP = "192.168.1.100"
        private const val DEFAULT_CONSOLE_PORT = 49280
        private const val DEFAULT_TESTING_IP = "192.168.1.200"
        private const val DEFAULT_TESTING_PORT = 8080
        private const val DEFAULT_TIMEOUT_SECONDS = 10
        private const val DEFAULT_ENABLE_LOGGING = true
    }
    
    /**
     * Network target types
     * Equivalent to iOS NetworkTargetType enum
     */
    @Serializable
    enum class NetworkTargetType(val displayName: String) {
        CONSOLE("Yamaha Console"),
        TESTING_GUI("Mac GUI");
        
        companion object {
            val allCases: List<NetworkTargetType> = values().toList()
        }
    }
    
    /**
     * Connection status types
     * Equivalent to iOS ConnectionStatus enum
     */
    @Serializable
    enum class ConnectionStatus(val displayName: String) {
        DISCONNECTED("Disconnected"),
        CONNECTING("Connecting..."),
        CONNECTED("Connected"),
        ERROR("Connection Error");
        
        val isConnected: Boolean
            get() = this == CONNECTED
        
        val isError: Boolean
            get() = this == ERROR
    }
    
    // State flows for reactive UI updates
    private val _consoleIP = MutableStateFlow(loadConsoleIP())
    val consoleIPFlow: StateFlow<String> = _consoleIP.asStateFlow()
    
    private val _consolePort = MutableStateFlow(loadConsolePort())
    val consolePortFlow: StateFlow<Int> = _consolePort.asStateFlow()
    
    private val _testingIP = MutableStateFlow(loadTestingIP())
    val testingIPFlow: StateFlow<String> = _testingIP.asStateFlow()
    
    private val _testingPort = MutableStateFlow(loadTestingPort())
    val testingPortFlow: StateFlow<Int> = _testingPort.asStateFlow()
    
    private val _targetType = MutableStateFlow(loadTargetType())
    val targetTypeFlow: StateFlow<NetworkTargetType> = _targetType.asStateFlow()
    
    private val _connectionStatus = MutableStateFlow(loadConnectionStatus())
    val connectionStatusFlow: StateFlow<ConnectionStatus> = _connectionStatus.asStateFlow()
    
    private val _timeoutSeconds = MutableStateFlow(loadTimeoutSeconds())
    val timeoutSecondsFlow: StateFlow<Int> = _timeoutSeconds.asStateFlow()
    
    private val _enableLogging = MutableStateFlow(loadEnableLogging())
    val enableLoggingFlow: StateFlow<Boolean> = _enableLogging.asStateFlow()
    
    private val _lastConnectionTime = MutableStateFlow(loadLastConnectionTime())
    val lastConnectionTimeFlow: StateFlow<Long?> = _lastConnectionTime.asStateFlow()
    
    init {
        if (BuildConfig.ENABLE_LOGGING) {
            Log.d(TAG, "🌐 NetworkSettings initialized")
            Log.d(TAG, "📊 Current settings: ${getCurrentSettingsSummary()}")
        }
    }
    
    /**
     * Console IP address
     * Equivalent to iOS consoleIP property
     */
    var consoleIP: String
        get() = _consoleIP.value
        set(value) {
            if (value != _consoleIP.value) {
                sharedPreferences.edit().putString(PREF_CONSOLE_IP, value).apply()
                _consoleIP.value = value
                
                if (BuildConfig.ENABLE_LOGGING) {
                    Log.d(TAG, "🖥️ Console IP updated: $value")
                }
            }
        }
    
    /**
     * Console port number
     * Equivalent to iOS consolePort property
     */
    var consolePort: Int
        get() = _consolePort.value
        set(value) {
            if (value != _consolePort.value) {
                sharedPreferences.edit().putInt(PREF_CONSOLE_PORT, value).apply()
                _consolePort.value = value
                
                if (BuildConfig.ENABLE_LOGGING) {
                    Log.d(TAG, "🔌 Console port updated: $value")
                }
            }
        }
    
    /**
     * Testing GUI IP address
     * Equivalent to iOS testingIP property
     */
    var testingIP: String
        get() = _testingIP.value
        set(value) {
            if (value != _testingIP.value) {
                sharedPreferences.edit().putString(PREF_TESTING_IP, value).apply()
                _testingIP.value = value
                
                if (BuildConfig.ENABLE_LOGGING) {
                    Log.d(TAG, "🧪 Testing IP updated: $value")
                }
            }
        }
    
    /**
     * Testing GUI port number
     * Equivalent to iOS testingPort property
     */
    var testingPort: Int
        get() = _testingPort.value
        set(value) {
            if (value != _testingPort.value) {
                sharedPreferences.edit().putInt(PREF_TESTING_PORT, value).apply()
                _testingPort.value = value
                
                if (BuildConfig.ENABLE_LOGGING) {
                    Log.d(TAG, "🔧 Testing port updated: $value")
                }
            }
        }
    
    /**
     * Target type selection
     * Equivalent to iOS targetType property
     */
    var targetType: NetworkTargetType
        get() = _targetType.value
        set(value) {
            if (value != _targetType.value) {
                sharedPreferences.edit().putString(PREF_TARGET_TYPE, value.name).apply()
                _targetType.value = value
                
                if (BuildConfig.ENABLE_LOGGING) {
                    Log.d(TAG, "🎯 Target type updated: ${value.displayName}")
                }
            }
        }
    
    /**
     * Connection status
     * Equivalent to iOS connectionStatus property
     */
    var connectionStatus: ConnectionStatus
        get() = _connectionStatus.value
        set(value) {
            if (value != _connectionStatus.value) {
                sharedPreferences.edit().putString(PREF_CONNECTION_STATUS, value.name).apply()
                _connectionStatus.value = value
                
                if (BuildConfig.ENABLE_LOGGING) {
                    Log.d(TAG, "🔗 Connection status updated: ${value.displayName}")
                }
            }
        }
    
    /**
     * Connection timeout in seconds
     * Equivalent to iOS timeoutSeconds property
     */
    var timeoutSeconds: Int
        get() = _timeoutSeconds.value
        set(value) {
            if (value != _timeoutSeconds.value && value > 0) {
                sharedPreferences.edit().putInt(PREF_TIMEOUT_SECONDS, value).apply()
                _timeoutSeconds.value = value
                
                if (BuildConfig.ENABLE_LOGGING) {
                    Log.d(TAG, "⏱️ Timeout updated: ${value}s")
                }
            }
        }
    
    /**
     * Enable network logging
     * Equivalent to iOS enableLogging property
     */
    var enableLogging: Boolean
        get() = _enableLogging.value
        set(value) {
            if (value != _enableLogging.value) {
                sharedPreferences.edit().putBoolean(PREF_ENABLE_LOGGING, value).apply()
                _enableLogging.value = value
                
                if (BuildConfig.ENABLE_LOGGING) {
                    Log.d(TAG, "📝 Logging ${if (value) "enabled" else "disabled"}")
                }
            }
        }
    
    /**
     * Last successful connection time
     * Equivalent to iOS lastConnectionTime property
     */
    var lastConnectionTime: Long?
        get() = _lastConnectionTime.value
        set(value) {
            val timestamp = value ?: 0L
            sharedPreferences.edit().putLong(PREF_LAST_CONNECTION_TIME, timestamp).apply()
            _lastConnectionTime.value = if (timestamp > 0L) timestamp else null
            
            if (BuildConfig.ENABLE_LOGGING && value != null) {
                Log.d(TAG, "⏰ Last connection time updated: ${java.util.Date(value)}")
            }
        }
    
    /**
     * Get current target IP based on target type
     * Equivalent to iOS currentTargetIP computed property
     */
    val currentTargetIP: String
        get() = when (targetType) {
            NetworkTargetType.CONSOLE -> consoleIP
            NetworkTargetType.TESTING_GUI -> testingIP
        }
    
    /**
     * Get current target port based on target type
     * Equivalent to iOS currentTargetPort computed property
     */
    val currentTargetPort: Int
        get() = when (targetType) {
            NetworkTargetType.CONSOLE -> consolePort
            NetworkTargetType.TESTING_GUI -> testingPort
        }
    
    /**
     * Get current target URL
     * Equivalent to iOS currentTargetURL computed property
     */
    val currentTargetURL: String
        get() = "http://$currentTargetIP:$currentTargetPort"
    
    /**
     * Check if current settings are valid
     * Equivalent to iOS isValidConfiguration computed property
     */
    val isValidConfiguration: Boolean
        get() = currentTargetIP.isNotBlank() && 
                currentTargetPort in 1..65535 &&
                timeoutSeconds > 0 &&
                isValidIPAddress(currentTargetIP)
    
    /**
     * Save all current settings to SharedPreferences
     * Equivalent to iOS saveSettings method
     */
    fun saveSettings() {
        sharedPreferences.edit().apply {
            putString(PREF_CONSOLE_IP, consoleIP)
            putInt(PREF_CONSOLE_PORT, consolePort)
            putString(PREF_TESTING_IP, testingIP)
            putInt(PREF_TESTING_PORT, testingPort)
            putString(PREF_TARGET_TYPE, targetType.name)
            putString(PREF_CONNECTION_STATUS, connectionStatus.name)
            putInt(PREF_TIMEOUT_SECONDS, timeoutSeconds)
            putBoolean(PREF_ENABLE_LOGGING, enableLogging)
            lastConnectionTime?.let { putLong(PREF_LAST_CONNECTION_TIME, it) }
        }.apply()
        
        if (BuildConfig.ENABLE_LOGGING) {
            Log.d(TAG, "💾 Settings saved: ${getCurrentSettingsSummary()}")
        }
    }
    
    /**
     * Reset settings to default values
     * Equivalent to iOS resetToDefaults method
     */
    fun resetToDefaults() {
        consoleIP = DEFAULT_CONSOLE_IP
        consolePort = DEFAULT_CONSOLE_PORT
        testingIP = DEFAULT_TESTING_IP
        testingPort = DEFAULT_TESTING_PORT
        targetType = NetworkTargetType.CONSOLE
        connectionStatus = ConnectionStatus.DISCONNECTED
        timeoutSeconds = DEFAULT_TIMEOUT_SECONDS
        enableLogging = DEFAULT_ENABLE_LOGGING
        lastConnectionTime = null
        
        saveSettings()
        
        if (BuildConfig.ENABLE_LOGGING) {
            Log.d(TAG, "🔄 Settings reset to defaults")
        }
    }
    
    /**
     * Get timeout in milliseconds
     * Equivalent to iOS timeoutMs computed property
     */
    val timeoutMs: Long
        get() = timeoutSeconds * 1000L
    
    /**
     * Validate IP address format
     */
    private fun isValidIPAddress(ip: String): Boolean {
        if (ip.isBlank()) return false
        
        val parts = ip.split(".")
        if (parts.size != 4) return false
        
        return parts.all { part ->
            try {
                val num = part.toInt()
                num in 0..255
            } catch (e: NumberFormatException) {
                false
            }
        }
    }
    
    /**
     * Get current settings summary for logging
     */
    private fun getCurrentSettingsSummary(): String {
        return "Target: ${targetType.displayName}, " +
               "IP: $currentTargetIP:$currentTargetPort, " +
               "Timeout: ${timeoutSeconds}s, " +
               "Status: ${connectionStatus.displayName}"
    }
    
    // Private getters for initialization
    private fun loadConsoleIP(): String {
        return sharedPreferences.getString(PREF_CONSOLE_IP, DEFAULT_CONSOLE_IP) ?: DEFAULT_CONSOLE_IP
    }
    
    private fun loadConsolePort(): Int {
        return sharedPreferences.getInt(PREF_CONSOLE_PORT, DEFAULT_CONSOLE_PORT)
    }
    
    private fun loadTestingIP(): String {
        return sharedPreferences.getString(PREF_TESTING_IP, DEFAULT_TESTING_IP) ?: DEFAULT_TESTING_IP
    }
    
    private fun loadTestingPort(): Int {
        return sharedPreferences.getInt(PREF_TESTING_PORT, DEFAULT_TESTING_PORT)
    }
    
    private fun loadTargetType(): NetworkTargetType {
        val typeName = sharedPreferences.getString(PREF_TARGET_TYPE, NetworkTargetType.CONSOLE.name)
        return try {
            NetworkTargetType.valueOf(typeName ?: NetworkTargetType.CONSOLE.name)
        } catch (e: IllegalArgumentException) {
            NetworkTargetType.CONSOLE
        }
    }
    
    private fun loadConnectionStatus(): ConnectionStatus {
        val statusName = sharedPreferences.getString(PREF_CONNECTION_STATUS, ConnectionStatus.DISCONNECTED.name)
        return try {
            ConnectionStatus.valueOf(statusName ?: ConnectionStatus.DISCONNECTED.name)
        } catch (e: IllegalArgumentException) {
            ConnectionStatus.DISCONNECTED
        }
    }
    
    private fun loadTimeoutSeconds(): Int {
        return sharedPreferences.getInt(PREF_TIMEOUT_SECONDS, DEFAULT_TIMEOUT_SECONDS)
    }
    
    private fun loadEnableLogging(): Boolean {
        return sharedPreferences.getBoolean(PREF_ENABLE_LOGGING, DEFAULT_ENABLE_LOGGING)
    }
    
    private fun loadLastConnectionTime(): Long? {
        val timestamp = sharedPreferences.getLong(PREF_LAST_CONNECTION_TIME, 0L)
        return if (timestamp > 0L) timestamp else null
    }
}